src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,885 @@
+
+			if (!task->retry_count)
+			{
+				/* first retry, need to allocate a counter */
+				MemoryContext taskContext = GetMemoryChunkContext(task);
+				task->retry_count = MemoryContextAlloc(taskContext, sizeof(int32));
+				*task->retry_count = 1;
+			}
+			else
+			{
+				(*task->retry_count)++;
+			}
+
+			if (*task->retry_count >= 3)
+			{
+				/* fail after 3 retries */
+				task->status = BACKGROUND_TASK_STATUS_ERROR;
+
+				/* when we error this task, we need to unschedule all dependant tasks */
+				UnscheduleDependantTasks(task->jobid, task->taskid);
+			}
+			else
+			{
+				task->status = BACKGROUND_TASK_STATUS_SCHEDULED;
+			}
+		}
+		UpdateBackgroundTask(task);
+		UpdateBackgroundJob(task->jobid);
+
+		dsm_detach(seg);
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		DeepFreeBackgroundTask(task);
+		task = NULL;
+	}
+
+	MemoryContextSwitchTo(oldContextPerJob);
+	MemoryContextDelete(perTaskContext);
+}
+
+
+static const char *
+error_severity(int elevel)
+{
+	const char *prefix;
+
+	switch (elevel)
+	{
+		case DEBUG1:
+		case DEBUG2:
+		case DEBUG3:
+		case DEBUG4:
+		case DEBUG5:
+		{
+			prefix = gettext_noop("DEBUG");
+			break;
+		}
+
+		case LOG:
+		case LOG_SERVER_ONLY:
+		{
+			prefix = gettext_noop("LOG");
+			break;
+		}
+
+		case INFO:
+		{
+			prefix = gettext_noop("INFO");
+			break;
+		}
+
+		case NOTICE:
+		{
+			prefix = gettext_noop("NOTICE");
+			break;
+		}
+
+		case WARNING:
+		case WARNING_CLIENT_ONLY:
+		{
+			prefix = gettext_noop("WARNING");
+			break;
+		}
+
+		case ERROR:
+		{
+			prefix = gettext_noop("ERROR");
+			break;
+		}
+
+		case FATAL:
+		{
+			prefix = gettext_noop("FATAL");
+			break;
+		}
+
+		case PANIC:
+		{
+			prefix = gettext_noop("PANIC");
+			break;
+		}
+
+		default:
+		{
+			prefix = "???";
+			break;
+		}
+	}
+
+	return prefix;
+}
+
+
+/*
+ * bgw_generate_returned_message -
+ *      generates the message to be inserted into the job_run_details table
+ *      first part is comming from error_severity (elog.c)
+ */
+static void
+bgw_generate_returned_message(StringInfoData *display_msg, ErrorData edata)
+{
+	const char *prefix = error_severity(edata.elevel);
+	appendStringInfo(display_msg, "%s: %s", prefix, edata.message);
+	if (edata.detail != NULL)
+	{
+		appendStringInfo(display_msg, "\nDETAIL: %s", edata.detail);
+	}
+
+	if (edata.hint != NULL)
+	{
+		appendStringInfo(display_msg, "\nHINT: %s", edata.hint);
+	}
+
+	if (edata.context != NULL)
+	{
+		appendStringInfo(display_msg, "\nCONTEXT: %s", edata.context);
+	}
+}
+
+
+static void
+ConsumeTaskWorkerOutput(shm_mq_handle *responseq, BackgroundTask *task, bool *hadError)
+{
+	/*
+	 * Message-parsing routines operate on a null-terminated StringInfo,
+	 * so we must construct one.
+	 */
+	StringInfoData msg = { 0 };
+	initStringInfo(&msg);
+
+	for (;;)
+	{
+		resetStringInfo(&msg);
+
+		/* Get next message. */
+		Size nbytes = 0;
+		void *data = NULL;
+		shm_mq_result res = shm_mq_receive(responseq, &nbytes, &data, false);
```

Comment:
For future reference, the last argument is quite critical, since this is currently a blocking operation. To support concurrent tasks, we will need to change the loop to be non-blocking.